pkg/metrics/tls.go

```diff
@@ -49,17 +61,30 @@ type TLSConfigurerImpl struct {
 	certDir           string
 	clientCAConfigMap string
 	clientCANamespace string
+	k8sWatcher        *k8scfgmap.ConfigMapWatcher
+
+	clientCAs       []*x509.Certificate
+	serverCerts     []tls.Certificate
+	tlsConfigHolder *certwatch.TLSConfigHolder
+
+	mutex sync.RWMutex
 }
 
 var _ TLSConfigurer = (*TLSConfigurerImpl)(nil)
 
 // NewTLSConfigurer creates a new TLS configurer.
-func NewTLSConfigurer(certDir string, clientCANamespace, clientCAConfigMap string) (TLSConfigurer, error) {
+func NewTLSConfigurer(certDir string, k8sClient kubernetes.Interface, clientCANamespace, clientCAConfigMap string) (TLSConfigurer, error) {
+	tlsRootConfig := verifier.DefaultTLSServerConfig(nil, nil)
+	tlsRootConfig.ClientAuth = tls.RequireAndVerifyClientCert
 	cfgr := &TLSConfigurerImpl{
 		certDir:           certDir,
 		clientCANamespace: clientCANamespace,
 		clientCAConfigMap: clientCAConfigMap,
+		tlsConfigHolder:   certwatch.NewTLSConfigHolder(tlsRootConfig),
 	}
+	cfgr.tlsConfigHolder.AddServerCertSource(&cfgr.serverCerts)
+	cfgr.tlsConfigHolder.AddClientCertSource(&cfgr.clientCAs)
+	cfgr.k8sWatcher = k8scfgmap.NewConfigMapWatcher(k8sClient, cfgr.updateClientCA)
```

Comment:
Why not start the watchers from the get go and get rid of `WatchForChanges`? We anyways take into account whether `secureMetricsEnabled` and it seems like the potential for bugs when somebody misses the `WatchForChanges` call (which I wouldn't expect to be exposed within the TLS configurer to begin with, I'd just expect this to give me TLS config to use, I don't necessarily care whether its watching or one-time reading it).

Reply:
Done. That allows us also to re-use an existing Interface for `TLSConfigurer`.